8273459: Update code segment alignment to 64 bytes

asgibbons opened the pull request:
Change the default code entry alignment to 64 bytes from 32 bytes.  This allows for maintaining proper 64-byte alignment of data within a code segment, which is required by several AVX-512 instructions.

I ran into this while implementing Base64 encoding and decoding.  Code segments which were allocated with the address mod 32 == 0 but with the address mod 64 != 0 would cause the align() macro to misalign.  This is because the align macro aligns to the size of the code segment and not the offset of the PC.  So align(64) would align the PC to a multiple of 64 bytes from the start of the segment, and not to a pure 64-byte boundary as requested.  Changing the alignment of the segment to 64 bytes fixes the issue.

I have not seen any measurable difference in either performance or memory usage with the tests I have run.

See [this ](https://mail.openjdk.java.net/pipermail/hotspot-dev/2021-August/054180.html) article for the discussion thread.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273459](https://bugs.openjdk.java.net/browse/JDK-8273459): Update code segment alignment to 64 bytes


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5547/head:pull/5547` \
`$ git checkout pull/5547`

Update a local copy of the PR: \
`$ git checkout pull/5547` \
`$ git pull https://git.openjdk.java.net/jdk pull/5547/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5547`

View PR using the GUI difftool: \
`$ git pr show -t 5547`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5547.diff">https://git.openjdk.java.net/jdk/pull/5547.diff</a>

</details>

jatin-bhateja commented:
How about setting dynamic alignment based on MaxVectorSize ? i.e. match the alignment to MaxVectorSize. This way we can save internal fragmentations over AVX2.
It could be done during VM Initialization similar to following code.
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/vm_version_x86.cpp#L1439

jatin-bhateja commented on src/hotspot/cpu/x86/globals_x86.hpp:
```diff
@@ -44,7 +44,7 @@ define_pd_global(uintx, CodeCacheSegmentSize,    64 COMPILER1_AND_COMPILER2_PRES
 // only assured that the entry instruction meets the 5 byte size requirement.
 #if COMPILER2_OR_JVMCI
-define_pd_global(intx, CodeEntryAlignment,       32);
+define_pd_global(intx, CodeEntryAlignment,       64);
```
How about setting dynamic alignment based on MaxVectorSize ? i.e. match the alignment to MaxVectorSize. This way we can save internal fragmentations over AVX2.
It could be done during VM Initialization similar to following code.
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/vm_version_x86.cpp#L1439

erikj79 commented on .gitignore:
```diff
@@ -16,5 +16,5 @@ NashornProfile.txt
 **/JTwork/**
 /src/utils/LogCompilation/target/
-/.project/
+**.project**
```
Changes to .gitignore should probably be made separately.

vnkozlov replied:
Yes. This change should be removed from patch.

vnkozlov reviewed (changes requested):
I share Dean's concern from the discussion before. `CodeEntryAlignment` is used in a lot of places and we have to be careful about changes to it.
I found only 7 cases with `align(64)`:
```
src/hotspot/cpu/x86//stubGenerator_x86_64.cpp:    __ align(64);
src/hotspot/cpu/x86//stubGenerator_x86_64.cpp:    __ align(64);
src/hotspot/cpu/x86//stubGenerator_x86_64.cpp:    __ align(64);
src/hotspot/cpu/x86//stubGenerator_x86_64.cpp:    __ align(64);
src/hotspot/cpu/x86//stubGenerator_x86_32.cpp:    __ align(64);
src/hotspot/cpu/x86//stubGenerator_x86_32.cpp:    __ align(64);
src/hotspot/cpu/x86//stubGenerator_x86_32.cpp:    __ align(64);
```
It does not justify such general changes.
May suggestion would be to add new `align64()` method to call pc() as you suggested in original proposal.

vnkozlov commented:
New function may also use `MaxVectorSize` as Jatin proposed if it helps.

asgibbons commented:
I think I have not made the point clearly enough.  The `align` function is used to manipulate the address bits for the byte following the `align()`.  This means that wherever the code is copied, the address of that byte should have the appropriate address bit configuration in the copy (as well as the original, of course).  Since the current implementation is using the base address of the allocated segment to determine alignment, the only way to ensure the proper bit configuration of the address is to ensure the base address of the newly-allocated segment is aligned identically to the original.

I believe this is entirely independent of `MaxVectorSize`, so I don't believe it's appropriate to use this value for address alignment.  Using `pc()` fixes the case in the source segment, but will break 50% of the time when the segment is copied with a `CodeEntryAlignment` of 32.

I think the bottom line is that `align()` is broken for any value greater than `CodeEntryAlignment`.  I can foresee a case where it may be beneficial (from an algorithm perspective) to have large alignment values, like align(256) to simplify pointer arithmetic (for example).  All of these proposed changes will not ensure this alignment when a segment is copied.

Perhaps the appropriate thing to do is to put an `assert()` in `align()` to fail if the requested alignment cannot be ensured?

IMHO, the "right" thing to do is to mark the bytes requiring address alignment and handle the cases on copy.  This would add significant complexity, however.

jatin-bhateja commented:
> I think I have not made the point clearly enough. The `align` function is used to manipulate the address bits for the byte following the `align()`. This means that wherever the code is copied, the address of that byte should have the appropriate address bit configuration in the copy (as well as the original, of course). Since the current implementation is using the base address of the allocated segment to determine alignment, the only way to ensure the proper bit configuration of the address is to ensure the base address of the newly-allocated segment is aligned identically to the original.
> 
> I believe this is entirely independent of `MaxVectorSize`, so I don't believe it's appropriate to use this value for address alignment. Using `pc()` fixes the case in the source segment, but will break 50% of the time when the segment is copied with a `CodeEntryAlignment` of 32.
> 
> I think the bottom line is that `align()` is broken for any value greater than `CodeEntryAlignment`. I can foresee a case where it may be beneficial (from an algorithm perspective) to have large alignment values, like align(256) to simplify pointer arithmetic (for example). All of these proposed changes will not ensure this alignment when a segment is copied.
> 
> Perhaps the appropriate thing to do is to put an `assert()` in `align()` to fail if the requested alignment cannot be ensured?
> 
> IMHO, the "right" thing to do is to mark the bytes requiring address alignment and handle the cases on copy. This would add significant complexity, however.

Following code suggests that instr start addresses always honor CodeEntryAlignment.
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/asm/codeBuffer.cpp#L114

Current value of CodeEntryAligment is 32, a 32 byte aligned address is inherently 8, 16 byte aligned but not vice-versa.

Setting this value to 64 will cover cases for 8,16, 32 and 64 byte alignment constraints in stubGenerator_64.cpp. 

Also there are several location in stubGenerator.cpp using __ align(CodeEntryAlignment) and suddenly it will also ensure 64 byte alignment and create internal fragmentation issue, also as Vliadimir pointed out there are handful location which needs 64 byte alignment.  My  suggestion was, if you are attempting it then its scope should be extended only for AVX512 and hence MaxVectorSize usage was suggested since problem of alignment will majorly surface for vector instructions and MaxVectorSize value can be set to 32 even on AVX512 targets, thus its a robust indicator of vector size and associated alignment constraints.

dean-long commented:
It sounds like this will new alignment requirement will only be needed for stubs (initially?), but as proposed it will affect all other types of CodeBlobs.  Just looking at the affect during startup, I saw padding for BufferBlobs go from 24 to 56, RuntimeBlobs go from 0 to 32 and 16 to 48, and nmethods go from 24 to 56.

I would like to suggest again, to use the actual alignment requirements of the CodeBuffer to determine the alignment of the CodeBlob.

> Perhaps the appropriate thing to do is to put an assert() in align() to fail if the requested alignment cannot be ensured?

I agree.

> IMHO, the "right" thing to do is to mark the bytes requiring address alignment and handle the cases on copy. This would add significant complexity, however.

I disagree.  Let's not mark individual bytes.  The call to align() is enough to allow us to record the maximum alignment required by the CodeBuffer, and the added complexity is not at the individual instruction copy, but just choosing the correct alignment value when creating the CodeBlob.  For example, use MAX2(codebuffer->required_alignment(), CodeEntryAlignment) in place of CodeEntryAlignment.

And for my own curiousity, I would like to hear from Intel what the expected affect on icache performance is from increasing the alignment of code.

sviswa7 commented:
@asgibbons To me Vladimir Kozlov's suggestion of adding a align64() method calling pc() as you originally proposed looks the best. It meets our purpose and is limited in scope.

vnkozlov commented:
I am back from vacation!

I want to point out that when we generate code for these stubs we don't move them in CodeCache (in contrast to compiled methods): https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/stubRoutines.cpp#L268

`BufferBlob::create()` allocates specified space (of size `code_size2`, for example) directly in CodeCache in `NonNMethod` section:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/codeBlob.cpp#L232
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/code/codeBlob.cpp#L272

Based on that, using `pc()` and new `align64()` should be fine for these few cases.

jatin-bhateja commented:
HI @asgibbons,
Alignment of JIT sequence for loops work under influence of -XX:OptoLoopAlignment flag. This is currently set to 16 may  be we can change this to 64 and study the effect on ICache as a separate patch as pointed by @dean-long.

For stubs a new runtime constant StubCodeEntryAlignment can be added which determines the start address alignment of stub BufferBlobs. This will limit the impact of changes also will prevent any unwanted fragmentation issues.

Your current patch with new align64 macro looks good,  it will also restrict the scope of changes.

vnkozlov commented on src/hotspot/cpu/x86/macroAssembler_x86.cpp:
```diff
@@ -1170,6 +1170,11 @@ void MacroAssembler::addpd(XMMRegister dst, AddressLiteral src) {
   }
 }
 
+// See 8273459.  Function for ensuring 64-byte alignment, intended for stubs only.
```
I suggest to extend comment to explain why it will not for nmethods - because they are copied when published.

vnkozlov reviewed (changes requested):
Current changes looks good. I have only 2 comments.

vnkozlov commented:
> HI @asgibbons, Alignment of JIT sequence for loops work under influence of -XX:OptoLoopAlignment flag. This is currently set to 16 may be we can change this to 64 and study the effect on ICache as a separate patch as pointed by @dean-long.

I would be very careful changing loop code alignment. You will introduce a lot of NOP instruction into code to alight it which will be executed. I am fine with experimenting (running SPEC benchmarks) with different `OptoLoopAlignment` values - may be on modern CPUs 16 bytes may not be optimal. But there were a lot of small loops in JAva code which will regress if pre-loop code has a lot of NOPs.

> 
> For stubs a new runtime constant StubCodeEntryAlignment can be added which determines the start address alignment of stub BufferBlobs. This will limit the impact of changes also will prevent any unwanted fragmentation issues.

May be but what benefit you can get with different alignment for stubs code?

> 
> Your current patch with new align64 macro looks good, it will also restrict the scope of changes.

asgibbons commented:
Thanks for the comments.  I reverted the .gitignore change and added comments as requested.  I also found a couple of unmodified align(64) which were changed.

As suggested, I added an assert in align() to flag spots where alignment cannot be ensured.

vnkozlov commented:
Good. Let me test it before you push.

jatin-bhateja commented:
.

> > HI @asgibbons, Alignment of JIT sequence for loops work under influence of -XX:OptoLoopAlignment flag. This is currently set to 16 may be we can change this to 64 and study the effect on ICache as a separate patch as pointed by @dean-long.
> 
> I would be very careful changing loop code alignment. You will introduce a lot of NOP instruction into code to alight it which will be executed. I am fine with experimenting (running SPEC benchmarks) with different `OptoLoopAlignment` values - may be on modern CPUs 16 bytes may not be optimal. But there are a lot of code with small loops in Java which will regress if pre-loop code has a lot of NOPs.
> 
> > For stubs a new runtime constant StubCodeEntryAlignment can be added which determines the start address alignment of stub BufferBlobs. This will limit the impact of changes also will prevent any unwanted fragmentation issues.
> 
> May be but what benefit you can get with different alignment for stubs code?
> 
It was an alternative to creating a different align64 macro in anticipation that we may introduce more stubs / code section constants which may give better performance if aligned to 64 byte boundary for wide vector targets like AVX512. New flag is specific to stubs thus has  restrictive scope but at the same time parameterizable and can have different values for targets. 

As you pointed out currently there are handful of places where guaranteed 64 byte alignment is desired so current approach looks good.

> > Your current patch with new align64 macro looks good, it will also restrict the scope of changes.

sviswa7 commented:
Looks good to me as well.

asgibbons commented:
Tested tier 1-3 Linux.  All pass.

vnkozlov reviewed (approved):
My testing passed.

sviswa7 reviewed (approved):
The updated patch looks good to me as well.

asgibbons commented:
/integrate

asgibbons commented:
@vnkozlov @jatin-bhateja @dean-long - Thank you all for the comments and testing.  Much appreciated.

sviswa7 commented:
/sponsor

DamonFool commented:
Minimal build fails after this PR.
Please see https://github.com/openjdk/jdk/pull/5764 .
Thanks.

sviswa7 commented:
/backport jdk17u-dev
